automates/text_reading/src/main/scala/org/clulab/grounding/WikiGrounder.scala

```diff
@@ -0,0 +1,168 @@
+package org.clulab.grounding
+
+import java.io.File
+import com.typesafe.config.{Config, ConfigFactory}
+import upickle.default._
+import org.apache.commons.text.similarity.LevenshteinDistance
+import org.clulab.odin.Mention
+
+import scala.collection.mutable.ArrayBuffer
+import scala.sys.process.Process
+import upickle.default.{ReadWriter, macroRW}
+import ai.lum.common.ConfigUtils._
+import org.clulab.aske.automates.apps.{ExtractAndAlign, JSONDocExporter}
+import org.clulab.embeddings.word2vec.Word2Vec
+import org.clulab.utils.FileUtils
+
+import scala.concurrent.duration.DurationInt
+
+// alt labels should be a seq of strings, not a string: alternativeLabel
+case class sparqlWikiResult(searchTerm: String, conceptID: String, conceptLabel: String, conceptDescription: Option[String], alternativeLabel: Option[String], subClassOf: Option[String], score: Option[Double], source: String = "Wikidata")
+
+object sparqlWikiResult {
+  implicit val rw: ReadWriter[sparqlWikiResult] = macroRW
+}
+
+case class WikiGrounding(variable: String, groundings: Seq[sparqlWikiResult])
+object WikiGrounding {
+  implicit val rw: ReadWriter[WikiGrounding] = macroRW
+}
+
+case class SeqOfWikiGroundings(wikiGroundings: Seq[WikiGrounding])
+
+object SeqOfWikiGroundings {
+  implicit val rw: ReadWriter[SeqOfWikiGroundings] = macroRW
+}
+
+object WikidataGrounder {
+
+  val config: Config = ConfigFactory.load()
+  val sparqlDir: String = config[String]("grounding.sparqlDir")
+  val stopWords = FileUtils.loadFromOneColumnTSV("src/main/resources/stopWords.tsv")
+  val exporter = JSONDocExporter()
+
+def groundTermsToWikidataRanked(variable: String, terms_with_underscores: Seq[String], sentence: Seq[String], w2v: Word2Vec, k: Int): Option[Seq[sparqlWikiResult]] = {
+
+  val cacheFilePath: String = config[String]("grounding.WikiCacheFilePath")
+  val file = new File(cacheFilePath)
+  val fileCache = if (file.exists()) {
+    ujson.read(file)
+  } else ujson.Obj()
+
+  if (terms_with_underscores.nonEmpty) {
+    val terms = terms_with_underscores.map(_.replace("_", " "))
+    val resultsFromAllTerms = new ArrayBuffer[sparqlWikiResult]()
+
+    for (term <- terms) {
+      val term_list = terms.filter(_==term)
+      val result = new ArrayBuffer[String]()
+
+      if (fileCache.obj.contains(term) && fileCache.obj(term).str.nonEmpty) {
+       result.append(fileCache.obj(term).str)
+      } else {
+        val res = WikidataGrounder.runSparqlQuery(term, WikidataGrounder.sparqlDir)
+        result.append(res)
+      }
+      fileCache(term) = result.head
+      val allSparqlWikiResults = new ArrayBuffer[sparqlWikiResult]()
+      if (result.head.nonEmpty) {
+        val lineResults = new ArrayBuffer[sparqlWikiResult]()
+        val resultLines = result.head.split("\n")
+        for (line <- resultLines) {
+          val splitLine = line.trim().split("\t")
```

Comment:
@maxaalexeeva consider refactoring lines 72 -- 77 by using [iterable unpacking](https://riptutorial.com/scala/example/3070/tuple-extractors). That will likely make this code look a lot cleaner and more condensed. Plus it looks extra cool 😎

Reply:
I might keep it as is for now because this way it might be easier to keep track of which information is associated with which split line item and I anticipate the need to add more information from wikidata (all this assuming I understood the link correctly).